src/combinators/mod.rs

```diff
@@ -0,0 +1,135 @@
+                    match a_value {
+                        Ok(a_value) => {
+                            if ab.0.replace(a_value).is_none() {
+                                count += 1;
+                            }
+                        },
+                        Err(_) => { return ; }
+                    }
+                }
+                b_value = b.next() => {
+                    match b_value {
+                        Ok(b_value) => {
+                            if ab.1.replace(b_value).is_none() {
+                                count += 1;
+                            }
+                        },
+                        Err(_) => {
+                            return;
+                        }
+                    }
+                }
+            }
+            if count == 2 {
+                break (ab.0.unwrap(), ab.1.unwrap());
+            }
+        };
+        loop {
+            writer.write(ab.clone());
+
+            select! {
+                a_value = a.next() => {
+                    match a_value {
+                        Ok(a_value) => { ab.0 = a_value; }
+                        Err(_) => { return; }
+                    }
+                }
+                b_value = b.next() => {
+                    match b_value {
+                        Ok(b_value) => { ab.1 = b_value; }
+                        Err(_) => { return; }
+                    }
+                }
+            }
+        }
+    })
+}
+
+pub fn throttle<E>(read: E, duration: Duration) -> Eventual<E::Output>
+where
+    E: IntoReader,
+{
+    let mut read = read.into_reader();
+
+    Eventual::spawn(move |mut writer| async move {
+        while let Ok(mut next) = read.next().await {
+            let end = time::Instant::now() + duration;
+            loop {
+                // Allow replacing the value until the time is up. This
+                // necessarily introduces latency but de-duplicates when there
+                // are intermittent bursts. Not sure what is better. Matching
+                // common-ts for now.
+                select! {
+                    n = read.next() => {
+                        if let Ok(n) = n {
+                            next = n;
+                        } else {
+                            return;
+                        }
+                    }
+                    _ = time::sleep_until(end) => {
+                        break;
+                    }
+                }
+            }
+            writer.write(next);
+        }
+    })
+}
+
+// TODO: Add pipe? The "GC" semantics make this unclear. The idea
+// behind pipe is to produce some side effect, which is a desirable
+// end goal for eventuals (eg: pipe this value into a UI, or log the latest)
+// but the part that is not clear here is what to do when the UI goes out of
+// scope. Should pipe provide an explicit handle that cancels on drop?
+//
+// TODO: Probably do not add filter or reduce, they don't make as much sense for eventuals.
```

Comment:
The only use of `reduce` that I could find in the Gateway was a poor implementation of `log_errors`.

For example this one: [https://github.com/edgeandnode/gateway/blob/680d77b2f602addd243085b996a242b26364fb70/packages/gateway/src/agent-syncing/syncing-client.ts#L55-L96](https://github.com/edgeandnode/gateway/blob/680d77b2f602addd243085b996a242b26364fb70/packages/gateway/src/agent-syncing/syncing-client.ts#L55-L96)

The basic pattern here was to (ab)use reduce to get access to the previous value so it could be yielded again when an error was logged. As far as I could tell, reduce was used for no other purpose.

A better version of `log_errors` isn't bespoke for every case but is generalized like so:
```rust
fn log_errors(logger: Logger, source: Eventual<Result<T, Err>>) -> Eventual<T> {
    Eventual::spawn(move |writer| async move {
        loop {
            match source.next().await {
                Ok(Ok(v)) => out.write(v),
                Ok(Err(e)) => error!(logger, e),
                Err(_) => break;
            }
        }
    })
}
```
Then you have: `timer(interval).map(poll_for_changes).log_errors()` and it's done. The north star of this library is to get all the complexity out of the consumer code.

Stepping back from the use of reduce in The Gatway and addressing whether reduce has any use at all - I think it's tempting to think of Eventuals like Stream and apply all the commonly known operations to them (reduce included). This is a mistake in kind. An Eventual is an eventually consistent value which updates _itself_. Put another way, it is a subset of snapshots in time of a value. A Stream is a view of a progressively available ordered multi-set, where the values represent distinct items. There's a lot of libraries that deal with streams, and none other that I know of that deal with Eventuals. Reduce can be used to aggregate distinct values. But what does it mean to aggregate a subset of snapshots over time? I can't think of any compelling use-case for that.

If you do need access to previous values for aggregation, the above loop used in `log_errors` does get you that.

Reply:
This does remind me that we want an eventual-aware `retry` though. The difference between a "regular" retry is that the eventual one would prefer to use a new value if available over retrying an old value. I've added a comment to the repo.

Reply:
I tried to get `log_errors` work like this:
```rust
pub fn handle_errors<E, F, T, Error>(source: E, f: F) -> Eventual<T>
where
    E: IntoReader<Output = Result<T, Error>>,
    F: 'static + Send + Fn(Error),
    T: Value + std::fmt::Debug,
    Error: Value + std::fmt::Debug,
{
    let mut reader = source.into_reader();

    Eventual::spawn(move |mut writer| async move {
        loop {
            match dbg!(reader.next().await) {
                Ok(Ok(v)) => writer.write(v),
                Ok(Err(e)) => f(e),
                Err(_) => break,
            }
        }
    })
}
```

with the following test code:
```rust
use eventuals::*;
use tokio::test;
use std::future;
use std::sync::{Arc, Mutex};

#[test]
async fn basic() {
    let (mut writer, numbers) = Eventual::<u32>::new();

    let errors = Arc::new(Mutex::new(vec![]));
    let errors_writer = errors.clone();

    let validated: Eventual<Result<u32, u32>> = numbers.map(|n| match n % 2 {
        0 => future::ready(Ok(n)),
        _ => future::ready(Err(n)),
    });

    let mut even_numbers = handle_errors(validated, move |err: u32| {
        println!("Err: {}", err);
        let mut errors = errors_writer.lock().unwrap();
        errors.push(err.clone());
    }).subscribe();

    for n in [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10] {
        writer.write(n);
        assert_eq!(even_numbers.next().await.ok().unwrap(), n);
    }

    assert_eq!(*errors.lock().unwrap(), vec![1, 3, 5, 7, 9]);
}
```
That of course freezes after the first odd number is written, because there is no next value yet.

This exercise made me realize two things:

1. Implementing `log_errors` or `handle_errors`, or whatever we want to call it, is difficult enough that we should probably provide something. The above code doesn't account for the fact that most error types are not `'static + Send + Clone + Eq`. So it wouldn't even work with most fallible functions.

2. We don't have an equivalent of `await eventual.value()` that returns the latest value _or_ blocks, do we? Because `eventual.next().await` blocks if there has been no new value since the last call to `next()`.

Reply:
1.
The `handle_errors` thing you wrote seems nice to me. (Submit a PR?) I like the idea here of providing some handler. The only reason I hadn't supplied `log_errors` was to avoid taking a dependency on `Logger` which your idea elegantly avoids.

The reason for the "deadlock" is that you would normally need to spawn the writer. It's not particularly a fault of `handle_errors` but rather of the circular dependency that exists in the test code (write depending on read of same eventual). This code is wrong anyway, because it would have panicked. With the circular dependency you can still get this to work by adding `if n % 2 == 0 assert_eq!(even_numbers.next()..., n)` I think that most code will naturally avoid circular dependencies, but it's something to be aware of in test code in particular.

You can make almost any Error type impl `Value` by wrapping it in `ByAddress(Arc)` as long as they already impl `Send`. Most complex types (eg: `Vec<Indexer>`) will want to be wrapped in `ByAddress(Arc)` which will make clone and comparisons effectively free.

2.
We don't have that specifically. I think you can do `eventual.subscribe().next()` which works the same way, but may make the borrow checker unhappy because of the temporary `subscribe`. To get it to compile you may need the unfortunate...
```rust
let next = {
  let subscribe = eventual.subscribe();
  subscribe.await
};
```
This is sad enough that I think I'll go ahead and add `eventual.next` because that was very useful in the Gateway.

Reply:
Huh, `subscribe().next()` already works as a replacement for `value` (no complaint about the temporary from the borrow checker). So you could change your test to...

`assert_eq!(even_numbers.subscribe().next().await.ok().unwrap(), (n / 2) * 2);`
If you also remove the earlier `subscribe()`. I still think there is a case for `.value()`.